internal/pkg/deploy/upload/asset/asset.go

```diff
@@ -0,0 +1,117 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package customresource provides functionality to upload Copilot custom resources.
+package asset
+
+import (
+	"fmt"
+	"io"
+	"path/filepath"
+
+	"github.com/spf13/afero"
+)
+
+type filterAction int
+
+const (
+	include filterAction = iota + 1
+	exclude
+)
+
+// UploadFunc is the function signature to upload contents to a destination.
+type UploadFunc func(dest string, contents io.Reader) (url string, err error)
+
+// UploadInput is the input of Upload.
+type UploadInput struct {
+	Source      string
+	Destination string
+	Includes    []string
+	Excludes    []string
+	Upload      UploadFunc
+	Reader      *afero.Afero
+}
+
+// Upload uploads static assets to Cloud Storage.
+func Upload(in *UploadInput) ([]string, error) {
+	files, err := listFiles(in.Reader, in.Source)
+	if err != nil {
+		return nil, err
+	}
+	filter := filter{buildPatterns(in.Includes, in.Excludes)}
+	filteredFiles, err := filter.apply(files)
+	if err != nil {
+		return nil, err
+	}
+	// TODO: read file and upload. Remove file names from return.
+	return filteredFiles, nil
+}
+
+func buildPatterns(includes, excludes []string) []pattern {
+	var filterPatterns []pattern
+	// Make sure exclude patterns are applied before include patterns.
+	for _, syntax := range excludes {
+		filterPatterns = append(filterPatterns, pattern{
+			action: exclude,
+			syntax: syntax,
+		})
+	}
+	for _, syntax := range includes {
+		filterPatterns = append(filterPatterns, pattern{
+			action: include,
+			syntax: syntax,
+		})
+	}
+	return filterPatterns
+}
+
+func listFiles(fs *afero.Afero, path string) ([]string, error) {
```

Comment:
can we leverage https://pkg.go.dev/github.com/spf13/afero#Walk here? 

I'm thinking that we can create a fn that returns a https://pkg.go.dev/path/filepath#WalkFunc:
```go
func walkFnWithMatcher(fpaths []string, matcher filepathMatcher) filepath.WalkFunc {
   return func(path string, info fs.FileInfo, err error) error {
      if info.IsDir() {  return filepath.SkipDir }
      ok, err := matcher.match(path)
      if err != nil { return err }
      if ok { fpaths = append(fpaths, path) }
      return nil
   }
}
```

Something like this, my only concern is that I'm not sure if the append will work